### PR TITLE
Change the definition of PID parameters in sdf

### PR DESCRIPTION
**Pull request by maxehrhardt**
Position and velocity PIDs can now be created for all joints, or groups of joints.

For an example of sdf definition, also pull the newest version of rover_config.

### CODE REVIEW

File: include/rover_gazebo_plugins/rover_gazebo_joint_plugin.hpp [hunk 1]
```diff
@@ -6,6 +6,41 @@
...
+
+        <!-- Add a namespace -->
+        <ros>
+           <namespace>/my_namespace</namespace>
+        </ros>
+
+        <!-- Update rate in Hz -->
+        <update_rate>100.0</update_rate>
+        
+        <!-- Parameters for position PIDs -->
```
Comment (mvoellmy) [inline include/rover_gazebo_plugins/rover_gazebo_joint_plugin.hpp:30]: very nice example!

File: src/rover_gazebo_joint_plugin.cpp [hunk 2]
```diff
@@ -16,197 +16,224 @@
...
-    impl_->joint_controller_->Reset();
-}
+        // Load all the parameters from the sdf
+        impl_->LoadPIDParametersFromSDF(_sdf);
 
-void RoverGazeboJointPluginPrivate::OnUpdate(const gazebo::common::UpdateInfo &_info)
-{
-    std::lock_guard<std::mutex> lock(lock_);
+        // Set the PIDs of the joints accordingly to the given parameters
+        for (auto const &pair : impl_->position_pid_parameters)
```
Comment (mvoellmy) [inline src/rover_gazebo_joint_plugin.cpp:125]: why is this called pair?

File: src/rover_gazebo_joint_plugin.cpp [hunk 3]
```diff
@@ -16,197 +16,224 @@
...
+        {
+            auto pid_identifier = pair.first;
+            auto parameters = pair.second;
 
-    // Check for update constraints
-    double seconds_since_last_update = (_info.simTime - last_update_time_).Double();
-    if (seconds_since_last_update < update_period_)
+            RCLCPP_DEBUG(impl_->ros_node_->get_logger(), "Position PID parameter: %s, Values: %f %f %f", pid_identifier.c_str(), parameters.X(), parameters.Y(), parameters.Z());
+
+            for (auto const &pair : impl_->joint_controller_->GetJoints())
```
Comment (mvoellmy) [inline src/rover_gazebo_joint_plugin.cpp:132]: Might be safer to use a different name for this pair as we still have the pair of the pid_parameters defined.

File: src/rover_gazebo_joint_plugin.cpp [hunk 4]
```diff
@@ -16,197 +16,224 @@
...
+        }
+
+        for (auto const &pair : impl_->velocity_pid_parameters)
+        {
+            auto pid_identifier = pair.first;
+            auto parameters = pair.second;
+
+            RCLCPP_DEBUG(impl_->ros_node_->get_logger(), "Velocity PID parameter: %s, Values: %f %f %f", pid_identifier.c_str(), parameters.X(), parameters.Y(), parameters.Z());
+
+            for (auto const &pair : impl_->joint_controller_->GetJoints())
```
Comment (mvoellmy) [inline src/rover_gazebo_joint_plugin.cpp:153]: same as above

File: src/rover_gazebo_joint_plugin.cpp [hunk 5]
```diff
@@ -16,197 +16,224 @@
...
+        // Iterate over received joint command array and set controller targets
+        for (auto const cmd : _msg->joint_command_array)
         {
-            if (!joint_controller_->SetPositionTarget(model_->GetJoint(cmd.name)->GetScopedName(), cmd.value))
+            if (cmd.mode == "POSITION")
             {
-                RCLCPP_WARN(ros_node_->get_logger(), "Joint %s from recieved command was npt found in model.", cmd.name.c_str());
+                if (!joint_controller_->SetPositionTarget(model_->GetJoint(cmd.name)->GetScopedName(), cmd.value))
+                {
+                    RCLCPP_WARN(ros_node_->get_logger(), "Joint %s from recieved command was npt found in model.", cmd.name.c_str());
```
Comment (mvoellmy) [inline src/rover_gazebo_joint_plugin.cpp:221]: typo:
"Joint from **received** command was **not** found in model.

**Comment by wieset** [comment]
If I understand this correctly, this currently looks for matching substrings in joint names to set the PIDs for groups of joints. This could be problematic if people are not careful when choosing joint names. Random example:

STR_WEAK, STR_STRONG, DRV_WEAK, DRV_STRONG

For STR this would match three joints.

Preferrably it would be good to be able to define joint groups. Then you could refer to them and expand them to a list of joints. Do you think that would be possible? Something like this:

```
<joint_group name="STR">
  <joint_name>joint_STR_LF</joint_name>
  <joint_name>joint_STR_RF</joint_name>
  <joint_name>joint_STR_LM</joint_name>
  <joint_name>joint_STR_RM</joint_name>
  <joint_name>joint_STR_LR</joint_name>
  <joint_name>joint_STR_RR</joint_name>
</joint_group>

<position_pids>
  <values group="STR">20.0 0.1 0.0</values>
<position_pids>
```

**Comment by wieset** [comment]
Also, `exoter.gazebo` in rover_config needs to be updated to the new format.

File: src/rover_gazebo_joint_plugin.cpp [hunk 6]
```diff
@@ -136,17 +137,24 @@ namespace gazebo_plugins
...
+                    parameter_has_joint = true;
                     impl_->joint_controller_->SetPositionPID(joint_name, gazebo::common::PID(parameters.X(), parameters.Y(), parameters.Z()));
                     impl_->joint_controller_->SetPositionTarget(joint_name, 0.0);
                     RCLCPP_DEBUG(impl_->ros_node_->get_logger(), "Set position PID on joint: %s to P: %f I: %f D: %f", joint_name.c_str(), parameters.X(), parameters.Y(), parameters.Z());
                 }
             }
+
+            if(!parameter_has_joint)
+            {
+                RCLCPP_WARN(impl_->ros_node_->get_logger(), "No joint found for Parameter %s. Maybe there is a type or your config is not up to date.", pid_identifier.c_str());
```
Comment (wieset) [inline src/rover_gazebo_joint_plugin.cpp:149]: Typo in "typo" :)

File: src/rover_gazebo_joint_plugin.cpp [hunk 7]
```diff
@@ -157,11 +165,16 @@ namespace gazebo_plugins
...
                 {
+                    parameter_has_joint = true;
                     impl_->joint_controller_->SetVelocityPID(joint_name, gazebo::common::PID(parameters.X(), parameters.Y(), parameters.Z()));
                     impl_->joint_controller_->SetVelocityTarget(joint_name, 0.0);
                     RCLCPP_DEBUG(impl_->ros_node_->get_logger(), "Set velocity PID on joint: %s to P: %f I: %f D: %f", joint_name.c_str(), parameters.X(), parameters.Y(), parameters.Z());
                 }
             }
+            if(!parameter_has_joint)
+            {
+                RCLCPP_WARN(impl_->ros_node_->get_logger(), "No joint found for Parameter %s. Maybe there is a type or your config is not up to date.", pid_identifier.c_str());
```
Comment (wieset) [inline src/rover_gazebo_joint_plugin.cpp:176]: Typo in "typo" :)

File: src/rover_gazebo_joint_plugin.cpp (hunk 6 above)
Comment (maxehrhardt) [reply src/rover_gazebo_joint_plugin.cpp:149]: Thats a great one

**Pull request merged**